test/hotspot/jtreg/compiler/cha/TypeProfileFinalMethod.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2021, 2024, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2024, Alibaba Group Holding Limited. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @summary test c1 to record type profile with CHA optimization
+ * @requires (vm.opt.TieredStoAtLevel == null | vm.opt.TieredStopAtLevel == 4)
+ * @library /test/lib
+ * @run driver compiler.cha.TypeProfileFinalMethod
+ */
+package compiler.cha;
+
+import java.io.File;
+import java.util.regex.Pattern;
+import java.util.regex.Matcher;
+import jdk.test.lib.Asserts;
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+
+public class TypeProfileFinalMethod {
+    public static void main(String[] args) throws Exception {
+        if (args.length == 1 && args[0].equals("Run")) {
```

Comment:
Instead of a check at runtime, you can introduce a separate class which drives test logic. Take a look at `compiler/jsr292/MHInlineTest.java` for an example (or grep for `class Launcher` under `test/hotspot/jtreg`).

Reply:
Fixed.